### PR TITLE
Map 104 avoid multiple requests

### PR DESCRIPTION
**Pull request by hasanrq**

### CODE REVIEW

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Controllers/DraftApprenticeshipControllerTests/WhenIEditADraftApprenticeship.cs [hunk 1]
```diff
@@ -1,7 +1,6 @@
-﻿using System.Threading.Tasks;
-using NUnit.Framework;
+﻿using NUnit.Framework;
 using SFA.DAS.CommitmentsV2.Api.Types.Validation;
-using SFA.DAS.ProviderCommitments.Web.Models.OveralppingTrainingDate;
+using System.Threading.Tasks;
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Controllers/DraftApprenticeshipControllerTests/WhenIEditADraftApprenticeship.cs:3]: system usings should come first. Change visual studio settings

Comment (hasanrq) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web.UnitTests/Controllers/DraftApprenticeshipControllerTests/WhenIEditADraftApprenticeship.cs]: Updated!

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs [hunk 2]
```diff
@@ -47,13 +47,59 @@ IFeatureTogglesService<ProviderFeatureToggle> featureTogglesService
...
         [Route("overlap-options")]
         public IActionResult DraftApprenticeshipOverlapOptions(DraftApprenticeshipOverlapOptionRequest request)
         {
             var apprenticeshipDetails = _commitmentsApiClient.GetApprenticeship(request.ApprenticeshipId.Value).Result;
-
             var featureToggleEnabled = _featureTogglesService.GetFeatureToggle(ProviderFeature.OverlappingTrainingDateWithoutPrefix).IsEnabled;
+
+            if (request.DraftApprenticeshipId.HasValue)
+            {
+                var pendingOverlapRequests = _outerApiService.GetOverlapRequest(request.DraftApprenticeshipId.Value).Result;
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs:85]: this should be awaited

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs [hunk 3]
```diff
@@ -47,13 +47,59 @@ IFeatureTogglesService<ProviderFeatureToggle> featureTogglesService
...
+        {
+            var mappedViewModel = await _modelMapper.Map<DraftApprenticeshipOverlapOptionViewModel>(viewModel);
+            return await DraftApprenticeshipOverlapOptions(mappedViewModel);
+        }
+
         [HttpGet]
         [Route("overlap-options")]
         public IActionResult DraftApprenticeshipOverlapOptions(DraftApprenticeshipOverlapOptionRequest request)
         {
             var apprenticeshipDetails = _commitmentsApiClient.GetApprenticeship(request.ApprenticeshipId.Value).Result;
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs:80]: this should be awaited

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs [hunk 4]
```diff
@@ -47,13 +47,59 @@ IFeatureTogglesService<ProviderFeatureToggle> featureTogglesService
...
+        [Route("overlap-options-with-pending-request")]
+        public IActionResult DraftApprenticeshipOverlapOptionsWithPendingRequest(DraftApprenticeshipOverlapOptionWithPendingRequest request)
+        {
+            var featureToggleEnabled = _featureTogglesService.GetFeatureToggle(ProviderFeature.OverlappingTrainingDateWithoutPrefix).IsEnabled;
+            var vm = new DraftApprenticeshipOverlapOptionWithPendingRequestViewModel()
+            {
+                CohortReference = request.CohortReference,
+                DraftApprenticeshipHashedId = request.DraftApprenticeshipHashedId,
+                DraftApprenticeshipId = request.DraftApprenticeshipId,
+                OverlappingTrainingDateRequestToggleEnabled = featureToggleEnabled,
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs:60]: this property isn't used in the view so why are we setting it in the viewmodel?

**Review by cofaulco — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs [hunk 5]
```diff
@@ -47,13 +47,59 @@ IFeatureTogglesService<ProviderFeatureToggle> featureTogglesService
...
+            };
+            return View(vm);
+        }
+
+        [HttpPost]
+        [Route("overlap-options-with-pending-request")]
+        public async Task<IActionResult> DraftApprenticeshipOverlapOptionsWithPendingRequest(DraftApprenticeshipOverlapOptionWithPendingRequestViewModel viewModel)
+        {
+            var mappedViewModel = await _modelMapper.Map<DraftApprenticeshipOverlapOptionViewModel>(viewModel);
+            return await DraftApprenticeshipOverlapOptions(mappedViewModel);
```
Comment (cofaulco) [inline src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs:73]: I would avoid calling another controller action from here. If there is shared logic then extract it to a private method and call from both

Comment (hasanrq) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs]: updated (for minimum code changes I just moved method body to private method)

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs (hunk 4 above)
Comment (hasanrq) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs]: Updated

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs (hunk 2 above)
Comment (hasanrq) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs]: updated

File: src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs (hunk 3 above)
Comment (hasanrq) [reply src/SFA.DAS.ProviderCommitments/SFA.DAS.ProviderCommitments.Web/Controllers/OverlappingTrainingDateRequestController.cs]: updated

**Review by cofaulco — APPROVED** [review]

**Comment by hasanrq** [comment]
updated with revised endpoint name

**Pull request merged**